src/ws/cockpitauthprocess.c

```diff
@@ -0,0 +1,738 @@
+  if (self->child_data.auth_fd > -1)
+    close (self->child_data.auth_fd);
+  self->child_data.auth_fd = -1;
+
+  g_free (self->id);
+  g_free (self->logname);
+  g_free (self->name);
+  G_OBJECT_CLASS (cockpit_auth_process_parent_class)->finalize (object);
+}
+
+static void
+spawn_child_setup (gpointer data)
+{
+  ChildFds *child_fds = data;
+  gint wanted;
+  gint large;
+  gint small;
+
+  if (child_fds->agent_fd > 0)
+    {
+      /* Two fds to keep open, close everything bigger than the larger
+       * of the two and then everything except smaller upto the larger fd
+       */
+
+      large = child_fds->auth_fd > child_fds->agent_fd ? child_fds->auth_fd : child_fds->agent_fd;
+      small = child_fds->auth_fd < child_fds->agent_fd ? child_fds->auth_fd : child_fds->agent_fd;
+
+      if (cockpit_unix_fd_close_all (large, large) < 0)
+        {
+          g_printerr ("couldn't close larger file descriptors: %m");
+          _exit (127);
+        }
+
+      if (cockpit_unix_fd_close_until (3, small, large) < 0)
+        {
+          g_printerr ("couldn't close smaller file descriptors: %m");
+          _exit (127);
+        }
+    }
+  else
+    {
+      if (cockpit_unix_fd_close_all (3, child_fds->auth_fd) < 0)
+        {
+          g_printerr ("couldn't close file descriptors: %m");
+          _exit (127);
+        }
+    }
+
+  /* Dup to the configured fd */
+  if (child_fds->auth_fd != child_fds->wanted_fd_number &&
+      dup2 (child_fds->auth_fd, child_fds->wanted_fd_number) < 0)
+    {
+      g_printerr ("couldn't dup file descriptor: %m");
+      _exit (127);
+    }
+
+  if (child_fds->auth_fd != child_fds->wanted_fd_number)
+    close (child_fds->auth_fd);
+
+  if (child_fds->agent_fd > 0)
+    {
+      /* Dup to the wanted fd + 1 */
+      wanted = child_fds->wanted_fd_number + 1;
+      if (dup2 (child_fds->agent_fd, wanted) < 0)
+        {
+          g_printerr ("couldn't dup agent file descriptor: %m");
+          _exit (127);
+        }
+
+      if (child_fds->agent_fd != wanted)
+        close (child_fds->agent_fd);
+    }
+}
+
+static gboolean
+on_timeout (gpointer user_data)
+{
+  CockpitAuthProcess *self = user_data;
+  if (!self->pipe_closed)
+      cockpit_pipe_close (self->pipe, "timeout");
+
+  return FALSE;
+}
+
+static void
+cockpit_auth_process_expect_response (CockpitAuthProcess *self)
+{
+  if (self->timeout)
+    g_source_remove (self->timeout);
+
+  self->timeout = g_timeout_add_seconds (self->max_wait_pipe,
+                                         on_timeout, self);
+  self->send_signal = TRUE;
+}
+
+static void
+report_message (CockpitAuthProcess *self,
+                GBytes *data)
+{
+  if (!self->send_signal)
+    {
+      g_debug ("%s: Dropping auth message, not expecting response", self->logname);
+      return;
+    }
+
+  g_object_ref (self);
+  g_debug ("%s: reporting message", self->logname);
+  self->send_signal = FALSE;
+
+  g_signal_emit (self, signals[MESSAGE], 0, data);
+
+  if (self->timeout)
```

Comment:
The same 'timeout' doesn't tell me what this timeout is for. Can we name it 'response_timeout'?